### PR TITLE
feat(core-magistrate-transactions): ensure unique genesisHash per bridgechain

### LINKED ISSUE #3191: [2.6][T3] Disallow duplicate genesisHash values for bridgechains
**Summary**  
Currently it's possible to register a bridgechain with a `genesisHash` value that already is registered by another bridgechain. For security reasons and to ensure uniqueness for each bridgechain, this behaviour should be disallowed.

https://dexplorer.ark.io/api/bridgechains
```
{
  "meta": {
    [...]
  },
  "data": [
    {
      "bridgechainId": 1,
      "businessId": 1,
      "name": "BridgechainTest",
      "seedNodes": ["1.1.1.1", "1.2.3.4"],
      "genesisHash": "c6512e5b7697cbedb0e1e5b250a983057c6c98cc48d64c1cfa475ca81f7518a7",
      "bridgechainRepository": "https://github.com/ArkEcosystem/core"
    },
    {
      "bridgechainId": 52,
      "businessId": 95,
      "name": "Lemii-Bridgechain-0005ccc9",
      "seedNodes": ["1.1.1.1", "1.2.3.4"],
      "genesisHash": "c6512e5b7697cbedb0e1e5b250a983057c6c98cc48d64c1cfa475ca81f7518a7",
      "bridgechainRepository": "https://www.lemii-bridgechain-0005ccc9.com/repository"
    },
    {
      "bridgechainId": 120,
      "businessId": 80,
      "name": "Lemii-Bridgechain-01efb1e7",
      "seedNodes": ["1.1.1.1", "1.2.3.4"],
      "genesisHash": "c6512e5b7697cbedb0e1e5b250a983057c6c98cc48d64c1cfa475ca81f7518a7",
      "bridgechainRepository": "https://www.lemii-bridgechain-01efb1e7.com/repository"
    },
    [...]
  ]
}

```
**Environment**  
2.6.0-next.4

### PR DESCRIPTION
**Pull request by Lemii**
## Summary
This PR changes the following:

- Implement validation that disallows duplicate genesisHash values for bridgechain registrations
- Updates existing functional test to comply with new validation
- Adds a new functional test that specifically checks for duplicate genesisHash values

To avoid multiple iterations over the same bridgechain array, I've refactored the check for duplicate bridgechain names to also accommodate for genesisHash validation.

For performance reasons a standard for loop is used. If a more clean and modern ES6 solution is preferred, we could go with something like the following as well:

```
        if (bridgechains) {
            Object.values(bridgechains).forEach(bridgechain => {
                if (bridgechain.bridgechainAsset.name === data.asset.bridgechainRegistration.name) {
                    throw new BridgechainAlreadyRegisteredError();
                }

                if (bridgechain.bridgechainAsset.genesisHash === data.asset.bridgechainRegistration.genesisHash) {
                    throw new GenesisHashAlreadyRegisteredError();
                }
            });
        }
```

Closes #3191 

## Checklist

- [ ] Documentation _(if necessary)_
- [X] Tests _(if necessary)_
- [X] Ready to be merged

### CODE REVIEW

File: packages/core-magistrate-transactions/src/handlers/bridgechain-registration.ts [hunk 1]
```diff
@@ -65,13 +70,20 @@ export class BridgechainRegistrationTransactionHandler extends MagistrateTransac
...
-            bridgechains &&
-            Object.values(bridgechains).some(bridgechain => {
-                return bridgechain.bridgechainAsset.name === data.asset.bridgechainRegistration.name;
-            })
-        ) {
-            throw new BridgechainAlreadyRegisteredError();
+        if (bridgechains) {
+            const bridgechainValues: IBridgechainWalletAttributes[] = Object.values(bridgechains);
+
+            for (let i = 0, len = bridgechainValues.length; i < len; i++) {
```
Comment (faustbrian) [inline packages/core-magistrate-transactions/src/handlers/bridgechain-registration.ts]:
```suggestion
            for (const bridgechain of bridgechainValues) {
```

Easier to read and there will never be an amount of bridgechains or businesses to iterate over where a raw `for` loop would have any significant benefits, especially with the v8 optimisations for `for of` loops.

**Comment by Lemii** [comment]
codeclimate keeps failing with: 

Function throwIfCannotBeApplied has a Cognitive Complexity of 11 (exceeds 5 allowed). Consider refactoring. 

```
    public async throwIfCannotBeApplied(
        transaction: Interfaces.ITransaction,
        wallet: State.IWallet,
        walletManager: State.IWalletManager,
    ): Promise<void> {
```

I suppose I should just ignore this, considering I did not make any changes to that part of the code?

**Comment by supaiku0** [comment]
> codeclimate keeps failing with:
> 
> Function throwIfCannotBeApplied has a Cognitive Complexity of 11 (exceeds 5 allowed). Consider refactoring.
> 
> ```
>     public async throwIfCannotBeApplied(
>         transaction: Interfaces.ITransaction,
>         wallet: State.IWallet,
>         walletManager: State.IWalletManager,
>     ): Promise<void> {
> ```
> 
> I suppose I should just ignore this, considering I did not make any changes to that part of the code?

Yes, that's out of scope.

---

I think checking whether a bridgechain name/genesis hash is unique or not has to be done globally, not locally per wallet.

I.e.
```typescript
for (const wallet of walletManager.getIndex("businesses").values())
    const bridgechains = wallet.getAttribute("business.bridgechains");
    if (...)
```

**Comment by Lemii** [comment]
@supaiku0 I've implemented your request. It will now loop over all wallets, therefore checking for duplicate genesisHashes globally.

The bridgechain name check has not been included in this global check and has been reverted to its original form where it only checks the local wallet. The reason for this is that 'there can be multiple bridgechains with the same name on different businesses, because any further validation will be performed in the market place', which was mentioned to me by dated (thanks).

The duplicate genesisHash functional test has been adjusted to have bridgechains created from different businesses, therefore being more thorough.

**Pull request merged**